dev/io.openliberty.microprofile.reactive.streams.operators30.internal/src/io/openliberty/microprofile/reactive/streams/operators30/spi/impl/LibertyReactiveStreamsEngineImpl.java

```diff
@@ -0,0 +1,112 @@
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutorService;
+
+import org.eclipse.microprofile.reactive.streams.operators.core.ReactiveStreamsEngineResolver;
+import org.eclipse.microprofile.reactive.streams.operators.core.ReactiveStreamsFactoryImpl;
+import org.eclipse.microprofile.reactive.streams.operators.spi.Graph;
+import org.eclipse.microprofile.reactive.streams.operators.spi.ReactiveStreamsEngine;
+import org.eclipse.microprofile.reactive.streams.operators.spi.ReactiveStreamsFactoryResolver;
+import org.eclipse.microprofile.reactive.streams.operators.spi.SubscriberWithCompletionStage;
+import org.eclipse.microprofile.reactive.streams.operators.spi.UnsupportedStageException;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.reactivestreams.Processor;
+import org.reactivestreams.Publisher;
+
+import com.ibm.wsspi.threadcontext.WSContextService;
+
+import io.smallrye.mutiny.jakarta.streams.Engine;
+
+@Component(name = "io.openliberty.microprofile.reactive.streams.operators30.spi.impl.LibertyReactiveStreamsEngineImpl", service = { ReactiveStreamsEngine.class }, property = { "service.vendor=IBM" }, immediate = true, configurationPolicy = ConfigurationPolicy.IGNORE)
+public class LibertyReactiveStreamsEngineImpl extends Engine implements ReactiveStreamsEngine {
+
+    private ExecutorService executorService;
+
+    private WSContextService wsContextService;
+
+    @Activate
+    public void activate() {
+        ReactiveStreamsFactoryResolver.setInstance(new ReactiveStreamsFactoryImpl());
+        ReactiveStreamsEngineResolver.setInstance(this);
+    }
+
+    @Deactivate
+    public void deactivate() {
+        ReactiveStreamsEngineResolver.setInstance(null);
+        ReactiveStreamsFactoryResolver.setInstance(null);
+    }
+
+    @Reference
+    public void setExecutorService(ExecutorService executorService) {
+        this.executorService = executorService;
+    }
+
+    @Reference
+    public void setContextService(WSContextService wsContextService) {
+        this.wsContextService = wsContextService;
+    }
+
+    /**
+     * Get the executor service for getting threads from.
+     *
+     * @return the executor
+     */
+    public ExecutorService getExecutor() {
+        return this.executorService;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public <T> CompletionStage<T> buildCompletion(Graph graph) throws UnsupportedStageException {
+
+        PrivilegedAction<CompletionStage<T>> action = new PrivilegedAction<CompletionStage<T>>() {
+            @Override
+            public CompletionStage<T> run() {
+                return LibertyReactiveStreamsEngineImpl.super.buildCompletion(graph);
+            }
+        };
```

Comment:
I'm not sure why we needed a `PrivilegedAction` here. It's possible that the newer mutiny implementation doesn't need one any more. Maybe we should raise an issue to check where we still need PrivilegedActions.

Reply:
good question

Reply:
https://github.com/OpenLiberty/open-liberty/issues/25450